ttnte/assemblers/matrix_assembler.py

```diff
@@ -267,11 +278,77 @@ def _build_fission(self, Intg_int):
         return self._flatten(
             tn.sparse_coo_tensor(indices, F.values(), size=shape).coalesce()
         )
+    
+
+    def _build_source_integral(self, Intg_int, J, J_det, R, dR):
+        #construct an array of all possible combinations of knot vectors 
+        #basis
+        p = self._patch.degree_u
+        q = self._patch.degree_v
+        coords = np.zeros((2,self._I1*self._I2*(p + 1)*(q + 1))) 
+        
+        i,j = 0, 0
+        #indexed so that the i*self._I2*(p+1)*(q+1)+j*(p+1)*(q+1)+k*(q+1)+l element is i,j,k,l
+        #could maybe use itertools for this
+        for xknot in self._xknot_intervals[:(self._I1 + 1)]: #check this is knot vector
+            
+            for yknot in self._yknot_intervals[:(self._I2 + 1)]:#check if need +1
+                k,l=0
+
+                for xtilda in np.linspace(xknot, self._xknot_intervals[j+1], q+1):
+                    coords[0,i*self._I2*(p+1)*(q+1)+j*(p+1)*(q+1)+k*(q+1):(q+1)] = xtilda
+                    k+=1
+
+                for ytilda in np.linspace(yknot, self._yknot_intervals[j+1], q+1):
+                    coords[1,i*self._I2*(p+1)*(q+1)+j*(p+1)*(q+1)+l::(q+1)] = ytilda
+                    l+=1
+
+                j+=1
+
+            i+=1
+
+
+        #construct source tensor
+        Q_min = self._source.evaluate_patch(self._p, coords=coords
+                                            ).reshape(1,1,1,self._xs_server.num_groups,
+                                                      self._I1, self._I2, p + 1, q + 1)
+        
+        Q = np.tile(Q_min, [4,1,1,1,1,1,1,1])
+
+        # Build components
+        JQ = ctg.einsum(
+            "abcd,c,d,qikgabcd->qikgabcd",
```

Comment:
Keep the comment format consistent. I prefer `# Comment` rather than `#comment`. Also this has definitely not been formatted by [black](https://github.com/psf/black). Run the following in the `ttnte` base directory
```shell
pip install -e ".[dev]"
pre-commit install
```